lib/datadog/appsec/instrumentation/gateway.rb

```diff
@@ -1,35 +1,23 @@
 # frozen_string_literal: true
 
+require_relative 'gateway/middleware'
+
 module Datadog
   module AppSec
     # Instrumentation for AppSec
     module Instrumentation
       # Instrumentation gateway implementation
       class Gateway
-        # Instrumentation gateway middleware
-        class Middleware
-          attr_reader :key, :block
-
-          def initialize(key, &block)
-            @key = key
-            @block = block
-          end
-
-          def call(stack, env)
-            @block.call(stack, env)
-          end
-        end
-
-        private_constant :Middleware
-
         def initialize
           @middlewares = Hash.new { |h, k| h[k] = [] }
+          @pushed_events = {}
         end
 
         def push(name, env, &block)
-          block ||= -> {}
+          @pushed_events[name] = true
```

Comment:
Very stupid question: Is this a possible memory leak concern? E.g. are the set of names fixed, or is it possible for them to accumulate over time?

Reply:
Not stupid questions. The list of events are controlled by us and pushed by us. The "names" we push are no subject to generation (they are static) and they deliberately triggered by the instrumentation code.

Considering our control over that - it should be safe. We can restrict it by the list of known or the count if you think it makes sense to do.

_P.S Currently we are pushing through gateway_

```
identity.set_user
rails.request.action
rack.request
rack.response
rack.request.body
graphql.multiplex
sinatra.request.dispatch
sinatra.request.routed
```

Reply:
Going to add the note and warning, thanks for advise 🤝